File: 1622. fancy-sequence.java
class Fancy {

    static final long MOD = 1000000007;
    List<Long> arr;
    long mul, add;

    public Fancy() {
        arr = new ArrayList<>();
        mul = 1;
        add = 0;
    }

    long modPow(long a, long b) {
        long res = 1;
        a %= MOD;

        while (b > 0) {
            if ((b & 1) == 1)
                res = (res * a) % MOD;

            a = (a * a) % MOD;
            b >>= 1;
        }

        return res;
    }

    long modInverse(long x) {
        return modPow(x, MOD - 2);
    }

    public void append(int val) {
        long stored = ((val - add + MOD) % MOD * modInverse(mul)) % MOD;
        arr.add(stored);
    }

    public void addAll(int inc) {
        add = (add + inc) % MOD;
    }

    public void multAll(int m) {
        mul = (mul * m) % MOD;
        add = (add * m) % MOD;
    }

    public int getIndex(int idx) {
        if (idx >= arr.size()) return -1;
        return (int)((arr.get(idx) * mul + add) % MOD);
    }
    
    public static void main(String[] args) {
        // Test case
        String[] operations = {"Fancy", "append", "addAll", "append", "multAll", "getIndex", "addAll", "append", "multAll", "getIndex", "getIndex", "getIndex"};
        int[][] values = {{}, {2}, {3}, {7}, {2}, {0}, {3}, {10}, {2}, {0}, {1}, {2}};
        int[] expected = {-1, -1, -1, -1, -1, 10, -1, -1, -1, 26, 34, 20};
        
        System.out.println("Test case for Fancy Sequence:");
        System.out.println("Operations: [Fancy, append, addAll, append, multAll, getIndex, addAll, append, multAll, getIndex, getIndex, getIndex]");
        System.out.println("Values: [[], [2], [3], [7], [2], [0], [3], [10], [2], [0], [1], [2]]");
        System.out.println("Expected: [null, null, null, null, null, 10, null, null, null, 26, 34, 20]");
        System.out.println();
        
        Fancy fancy = new Fancy();
        
        for (int i = 1; i < operations.length; i++) {
            String op = operations[i];
            int val = values[i].length > 0 ? values[i][0] : 0;
            
            if (op.equals("append")) {
                fancy.append(val);
                System.out.println(op + "(" + val + ") -> null");
            } else if (op.equals("addAll")) {
                fancy.addAll(val);
                System.out.println(op + "(" + val + ") -> null");
            } else if (op.equals("multAll")) {
                fancy.multAll(val);
                System.out.println(op + "(" + val + ") -> null");
            } else if (op.equals("getIndex")) {
                int result = fancy.getIndex(val);
                System.out.println(op + "(" + val + ") -> " + result + " (Expected: " + expected[i] + ")");
            }
        }
    }
}

File: 1622. fancy-sequence.py
class Fancy:

    MOD = 10**9 + 7

    def __init__(self):
        self.arr = []
        self.mul = 1
        self.add = 0

    def mod_pow(self, a, b):
        res = 1
        a %= self.MOD
        while b > 0:
            if b & 1:
                res = (res * a) % self.MOD
            a = (a * a) % self.MOD
            b >>= 1
        return res

    def mod_inverse(self, x):
        return self.mod_pow(x, self.MOD - 2)

    def append(self, val: int) -> None:
        stored = ((val - self.add) % self.MOD * self.mod_inverse(self.mul)) % self.MOD
        self.arr.append(stored)

    def addAll(self, inc: int) -> None:
        self.add = (self.add + inc) % self.MOD

    def multAll(self, m: int) -> None:
        self.mul = (self.mul * m) % self.MOD
        self.add = (self.add * m) % self.MOD

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.arr):
            return -1
        return (self.arr[idx] * self.mul + self.add) % self.MOD

if __name__ == "__main__":
    # Test case
    fancy = Fancy()
    operations = ["Fancy", "append", "addAll", "append", "multAll", "getIndex", "addAll", "append", "multAll", "getIndex", "getIndex", "getIndex"]
    values = [[], [2], [3], [7], [2], [0], [3], [10], [2], [0], [1], [2]]
    expected = [None, None, None, None, None, 10, None, None, None, 26, 34, 20]
    
    print("Test case for Fancy Sequence:")
    print("Operations:", operations)
    print("Values:", values)
    print("Expected:", expected)
    print()
    
    results = []
    fancy = Fancy()
    
    for i in range(1, len(operations)):
        op = operations[i]
        val = values[i]
        
        if op == "append":
            fancy.append(val[0])
            result = None
        elif op == "addAll":
            fancy.addAll(val[0])
            result = None
        elif op == "multAll":
            fancy.multAll(val[0])
            result = None
        elif op == "getIndex":
            result = fancy.getIndex(val[0])
        
        results.append(result)
        print(f"{op}{val} -> {result} (Expected: {expected[i]})")